crates/rethnet_eth/src/signature.rs

```diff
@@ -23,6 +23,17 @@ pub fn public_key_to_address(public_key: PublicKey) -> Address {
 }
 
 /// Converts a private to an address using the provided context.
+/// ```
+/// use std::str::FromStr;
+/// use rethnet_eth::{Address, signature::private_key_to_address};
+/// assert_eq!(
+///     private_key_to_address(
+///         &secp256k1::Secp256k1::signing_only(),
+///         "0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80",
+///     ).unwrap(),
+///     Address::from_str("0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266").unwrap(),
+/// );
+/// ```
```

Comment:
I thought of the assertion not as being a test but as a way to visually demonstrate to the reader what the expected output of the function is. I was going off of (admittedly vague) memories of seeing lots of assertions of return values in other rust docs i've been reading, but I see that in the link you provided most of the methods documented there include assertions on expected output values, just like I had remembered seeing "in the wild."

Looking at this again (and perhaps using some functionality I didn't know about when I first wrote this), the example could be improved by [Hiding portions of the example](https://doc.rust-lang.org/rustdoc/write-documentation/documentation-tests.html#hiding-portions-of-the-example):

```suggestion
/// ```
/// # use std::str::FromStr;
/// # use rethnet_eth::{Address, signature::private_key_to_address};
/// assert_eq!(
///     private_key_to_address(
///         &secp256k1::Secp256k1::signing_only(),
///         "0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80",
///     ).unwrap(),
///     Address::from_str("0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266").unwrap(),
/// );
/// ```
```

Reply:
I don't think hiding the `use` statements solves the problems I pointed out. I thought that was actually useful, as it illustrates how to use the API in a real-world scenario (all STL examples do the same too).

My problem is with showing that the API works using an assertion that an arbitrary private key string matches an arbitrary output address. That's something that only makes sense to us, within the context of a test where we hardcode them with some explanation.

As a user of an API, I expect that the function works. I just want to see how to use an API, with maybe some simple, clear `unwrap`s or `asserts` that illustrate the expected usage.

Reply:
As usual, I'm happy to defer to your judgement here, but I just wanted to point out that these aren't exactly arbitrary strings. This is a well known address, and a well known private key. It's the first of the 20 local accounts that come with the default-configured Hardhat Network, so it's displayed at the top of the standard output from `npx hardhat node`. And, we refer to that specific address from multiple locations in our documentation, and it's referred to by many other tutorials on the web, both of which you can see in this Google search result: https://www.google.com/search?q=tutorial+"0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266"

Reply:
I feel like this is degraded into nitpicking parts of the argument that are not important to the core, so I'll recite my stance:

> Adding examples to function documentation is encouraged, to help users understand how to use the API. To this end, they should be simple/minimal. doctests are merely needed to validate that the code in these examples compiles and runs. Documentations will however not be used to add or replace unit tests of functionality. That's what the unit test section inside `mod tests` is for.

The PR that you submitted was intended as and qualifies as adding a unit test to the documentation. My [suggestion above](https://github.com/NomicFoundation/hardhat/pull/4180#discussion_r1270893237) shows how it can be corrected to a minimal example.